### PR TITLE
feat: /store api product types

### PR DESCRIPTION
**Pull request by patrick-medusajs**
## What
Allow users to fetch ProductTypes from the storefront API.

## Why
This endpoint will allow developers to implement better faceted product search in Medusa without the need for search plugin. Developers will be able to use this to render refinement lists based on types, like this:
image

## How
Endpoint `GET /store/products/types` and `GET /store/product-types` (use [product types listing in admin](https://github.com/medusajs/medusa/blob/master/packages/medusa/src/api/routes/admin/products/list-types.ts) as reference)

Support added in @medusajs/medusa-js
Support added in medusa-react

## Testing
Similar automated tests as `GET /admin/products/types` and `GET /admin/product-types`

---

Resolves CORE-699

### CODE REVIEW

File: packages/medusa/src/api/routes/store/products/index.ts
Comment (olivermrbl) [inline packages/medusa/src/api/routes/store/products/index.ts:50]: todo: If the endpoint is not removed, I think we should change the response to `product_types` to eliminate ambiguity around what types, we are working with

**Review by olivermrbl — COMMENTED** [review]
Awesome work @patrick-medusajs - I've left a couple of comments, but no big changes required here 💪

**Review by adrien2p — COMMENTED** [review]
nice work 💪

File: packages/medusa/src/api/routes/store/products/list-types.ts
Comment (patrick-medusajs) [reply packages/medusa/src/api/routes/store/products/list-types.ts:1]: Removing the endpoint would introduce a breaking change in the API. If we wish to keep theses changes as a `patch` changelog, my recommendation would be to set `deprecated:true` on the OpenAPI Spec for this route and create a Linear ticket for it future removal.

If we don't mind the breaking change, then I'll mark this as a minor and proceed with removing the endpoint.

@olivermrbl waiting for your guidance before proceeding.

Comment (olivermrbl) [reply packages/medusa/src/api/routes/store/products/list-types.ts:1]: I am not entirely sure why removing the endpoint would introduce a breaking change, as it is not currently part of our store-facing API. Can I get you to elaborate?

It would introduce an inconsistency in our API as we would have `/admin/products/types` but not `/store/products/types`. So we could add the `deprecated` flag to `/admin/products/types` when we add `/admin/product-types`.

Let me know what you think. Also, sorry if I am missing something.

Comment (patrick-medusajs) [reply packages/medusa/src/api/routes/store/products/list-types.ts:1]: Ah yes, deprecate the `/admin` and not introduce the `/store`. I'll proceed with that. Thanks.

File: packages/medusa/src/api/routes/store/product-types/list-product-types.ts
Comment (patrick-medusajs) [reply packages/medusa/src/api/routes/store/product-types/list-product-types.ts:157]: Nice, I think all remove the definition StoreGetProductTypesPaginationParams and use FindPaginationParams directly on StoreGetProductTypesParams.
`export class StoreGetProductTypesParams extends FindPaginationParams`

The only side effect will be that the default limit will be 20 instead of 10 which I don't expect to be an issue.

I'll apply the same the equivalent Admin endpoint.

File: integration-tests/api/__tests__/store/product-type.js
Comment (patrick-medusajs) [reply integration-tests/api/__tests__/store/product-type.js:61]: I'll skip this one for now. I have yet to wrap my head on how I would like to move away from snapshots.

File: integration-tests/api/__tests__/store/product-type.js
Comment (patrick-medusajs) [reply integration-tests/api/__tests__/store/product-type.js:81]: I'll skip this one for now. I have yet to wrap my head on how I would like to move away from snapshots.

By the way, you're the one who create the test below on the Admin endpoint 🤣

File: packages/medusa/src/api/routes/store/products/list-types.ts
Comment (patrick-medusajs) [reply packages/medusa/src/api/routes/store/products/list-types.ts]: Done, see 237d679d36fcb6608fe2bcc689309454b77a9cbc

File: packages/medusa-js/src/resources/products.ts
Comment (patrick-medusajs) [reply packages/medusa-js/src/resources/products.ts]: Done, see dde73ab0468f30d08aca467221fb18a7255ba85f

File: packages/medusa/src/api/routes/store/products/index.ts
Comment (patrick-medusajs) [reply packages/medusa/src/api/routes/store/products/index.ts]: Done, see 0ee84569846c46b5fa54227249dbbafd10b87715

File: integration-tests/api/__tests__/store/product-type.js
Comment (patrick-medusajs) [reply integration-tests/api/__tests__/store/product-type.js]: Done, see 958700be25ca27896b043b455b09a966c7635524

File: integration-tests/api/__tests__/store/product-type.js
Comment (patrick-medusajs) [reply integration-tests/api/__tests__/store/product-type.js]: Done, see 958700be25ca27896b043b455b09a966c7635524

File: packages/medusa-js/src/resources/product-types.ts
Comment (patrick-medusajs) [reply packages/medusa-js/src/resources/product-types.ts]: Done, see dde73ab0468f30d08aca467221fb18a7255ba85f

File: packages/medusa-js/src/resources/product-types.ts
Comment (patrick-medusajs) [reply packages/medusa-js/src/resources/product-types.ts]: Done, see dde73ab0468f30d08aca467221fb18a7255ba85f

File: packages/medusa/src/api/routes/store/product-types/list-product-types.ts
Comment (patrick-medusajs) [reply packages/medusa/src/api/routes/store/product-types/list-product-types.ts]: Done, see 0ee84569846c46b5fa54227249dbbafd10b87715

File: integration-tests/api/__tests__/store/product-type.js
Comment (adrien2p) [reply integration-tests/api/__tests__/store/product-type.js:59]: It is just for consistency as most a the new tests are tested that way. Also, here what we really want to test is if the expected types are in the list (check id) and if the number of types is the one expected (here 2)

File: integration-tests/api/__tests__/store/product-type.js
Comment (adrien2p) [reply integration-tests/api/__tests__/store/product-type.js:75]: 😂 gosh, ahah thanks for reminding me. you could have take the compliment 👍 It is still a nice c/p ahah

File: packages/medusa/src/api/routes/store/product-types/list-product-types.ts
Comment (adrien2p) [reply packages/medusa/src/api/routes/store/product-types/list-product-types.ts]: yeah I don't think the default limit is a problem as it will be consistent when everything will extend it :)

File: packages/medusa-js/src/resources/admin/products.ts
Comment (olivermrbl) [inline packages/medusa-js/src/resources/admin/products.ts:71]: praise: Love this!

**Review by olivermrbl — COMMENTED** [review]
LGTM!

**Comment by olivermrbl** [comment]
@adrien2p Can I get you to give this another review?

**Comment by adrien2p** [comment]
> @adrien2p Can I get you to give this another review? 

@olivermrbl I am just waiting for patrick to answer my last answers but it is not blocking. Still, I would like to get his answer but otherwise it is good to go

File: integration-tests/api/__tests__/store/product-type.js
Comment (patrick-medusajs) [reply integration-tests/api/__tests__/store/product-type.js:59]: These where copied from the admin endpoint. I still stand by my decision to not refactor those tests.

In the new suggested format, I don't like that we assume that the seeded entities have id "XYZ". I rather be able to reference their value instead of assuming it explicitly. Something along the lines of:
```
import { mockProductTypes } from "product-seeded"
expect(res.data.product_types).toEqual(
  expect.arrayContainer([
    expect.objectContaining({
      id: mockProductTypes[0].id
    }),
    expect.objectContaining({
      id: mockProductTypes[1].id
    }),
  ])
)
```
I don't not want to further expand the scope of this PR.

Comment (adrien2p) [reply integration-tests/api/__tests__/store/product-type.js:59]: Sounds perfect, I agree with you

**Comment by patrick-medusajs** [comment]
@olivermrbl @adrien2p, is there anything missing before I clean-up my commits and merge? Still need that sweet green checkmark ✅.

**Review by adrien2p — APPROVED** [review]
LGTM 🚀

**Review by olivermrbl — APPROVED** [review]
LGTM! Good work Pat 🚀

**Comment by olivermrbl** [comment]
Feel free to merge as you please :)

**Pull request merged**
